graylog2-server/src/main/java/org/graylog2/cluster/leader/AutomaticLeaderElectionService.java

```diff
@@ -0,0 +1,147 @@
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.cluster.leader;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import org.graylog2.Configuration;
+import org.graylog2.cluster.lock.LockService;
+import org.graylog2.periodical.NodePingThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.Duration;
+
+@Singleton
+public class AutomaticLeaderElectionService extends AbstractExecutionThreadService implements LeaderElectionService {
+    private static final String RESOURCE_NAME = "cluster-leader";
+    private static final Logger log = LoggerFactory.getLogger(AutomaticLeaderElectionService.class);
+
+    public static final java.time.Duration DEFAULT_POLLING_INTERVAL = Duration.ofSeconds(2);
+
+    private final LockService lockService;
+    private final EventBus eventBus;
+    private final NodePingThread nodePingThread;
+    private final Duration pollingInterval;
+    private final Duration lockTTL;
+
+    private volatile boolean isLeader = false;
+    private Thread executionThread;
+    private long lastSuccess = 0;
+
+    @Inject
+    public AutomaticLeaderElectionService(Configuration configuration,
+                                          LockService lockService,
+                                          EventBus eventBus,
+                                          NodePingThread nodePingThread) {
+        this.pollingInterval = configuration.getLeaderElectionLockPollingInterval();
+        this.lockTTL = configuration.getLeaderElectionLockTTL();
+        this.lockService = lockService;
+        this.eventBus = eventBus;
+        this.nodePingThread = nodePingThread;
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        super.startUp();
+        executionThread = Thread.currentThread();
+    }
+
+    @Override
+    protected void triggerShutdown() {
+        executionThread.interrupt();
+    }
+
+    @Override
+    public boolean isLeader() {
+        return isLeader;
+    }
+
+    @Override
+    protected void run() throws Exception {
+        while (isRunning()) {
+            try {
+                runIteration();
+            } catch (Throwable t) {
+                log.error("Exception while acquiring/renewing leader lock.", t);
+            }
+        }
+    }
+
+    private void runIteration() {
+        final boolean wasLeader = isLeader;
+
+        try {
+            isLeader = lockService.lock(RESOURCE_NAME).isPresent();
+            lastSuccess = System.nanoTime();
+        } catch (Exception e) {
+            log.error("Unable to acquire/renew leader lock.", e);
+
+            final Duration timeSinceLastSuccess = Duration.ofNanos(System.nanoTime() - lastSuccess);
+            if (wasLeader && timeSinceLastSuccess.compareTo(lockTTL) >= 0) {
```

Comment:
Why do we wait for `lockTTL` on error instead of giving up the leader role immediately if locking throws an error?

Wouldn't it be safer to give up the leader role early?

Reply:
That's up to debate I think. We thought it would be beneficial not to give up the leader status upon first error and at least add some retries.
I'm OK with changing it though.

Reply:
Will be addressed in https://github.com/Graylog2/graylog2-server/pull/11713